### PR TITLE
Device List Filter - redo last filter

### PR DESCRIPTION
**Pull request by drcoolsanjeev**
### Type of change

- Enhancement / new feature

### Description

Modify implementation of last redo filter. Redo last filter will update the filter to the previous state.

### CODE REVIEW

**Comment by drcoolsanjeev** [comment]
It will be used once we implement container for this component.

**Comment by rkpattnaik780** [comment]
> @rkpattnaik780 `IDeviceFilterProp` will be used once we implement container for this component.

Okay, got it.

File: console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx [hunk 1]
```diff
@@ -123,21 +128,25 @@ const DeviceFilter: React.FunctionComponent<IDeviceFilterProps> = (
...
+    setFilter(getInitialFilter());
     setIsKebabOpen(false);
   };
   const onRedoFilter = () => {
     // redoFilter();
     const length = lastAppliedFilter.length;
-    const data = JSON.parse(
-      JSON.stringify({ ...lastAppliedFilter[length - 2] })
-    );
+    const data = createDeepCopy({ ...lastAppliedFilter[length - 2] });
```
Comment (anukritijha) [inline console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx:145]: Please rename data and datalist to be more descriptive.

File: console/console-init/ui/src/utils/utils.ts [hunk 2]
```diff
@@ -256,6 +256,10 @@ const getJson = (objects: any[]) => {
   return object;
 };
 
+const createDeepCopy = (object: any) => {
```
Comment (anukritijha) [inline console/console-init/ui/src/utils/utils.ts:259]: Again, please rename object to something descriptive.

File: console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx [hunk 3]
```diff
@@ -96,11 +95,12 @@ const DeviceFilter: React.FunctionComponent<IDeviceFilterProps> = (
...
 ) => {
-  const [filter, setFilter] = useState<IDeviceFilter>(setInitialFilter());
+  const [filter, setFilter] = useState<IDeviceFilter>(getInitialFilter());
   const [lastAppliedFilter, setLastAppliedFilter] = useState<IDeviceFilter[]>([
-    setInitialFilter()
+    getInitialFilter()
   ]);
   const [isKebabOpen, setIsKebabOpen] = useState<boolean>(false);
+  const [isRedoEnabled, setIsRedoEnabled] = useState<boolean>(false);
   const typeOptions: ISelectOption[] = [
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx:104]: can we move all dropdown options in constant file in same module. I would like to suggest create one constant file in utils at same module.

File: console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx [hunk 4]
```diff
@@ -111,6 +111,11 @@ const DeviceFilter: React.FunctionComponent<IDeviceFilterProps> = (
       key: "gateway",
       value: "gateway",
       label: "Using gateways"
+    },
+    {
+      key: "allTypes",
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx:116]: is it possible to make the key  in small case.

File: console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx [hunk 5]
```diff
@@ -123,21 +128,25 @@ const DeviceFilter: React.FunctionComponent<IDeviceFilterProps> = (
       key: "disabled",
       value: "disabled",
       label: "Disabled"
+    },
+    {
+      key: "allStatus",
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx:133]: is it possible to make the key in small case.

File: console/console-init/ui/src/utils/utils.ts (hunk 2 above)
Comment (ajaypratap003) [inline console/console-init/ui/src/utils/utils.ts:259]: i would like to suggest the name of function like deepCopy. what is your thought.

File: console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx (hunk 1 above)
Comment (drcoolsanjeev) [reply console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx]: done

File: console/console-init/ui/src/utils/utils.ts (hunk 2 above)
Comment (drcoolsanjeev) [reply console/console-init/ui/src/utils/utils.ts:259]: I think this is a generic function and name is descriptive as per the function

**Pull request merged**